src/main/java/org/terracotta/build/plugins/ToolPlugin.java

```diff
@@ -0,0 +1,374 @@
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.attribute.PosixFileAttributeView;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.stream.Stream;
+
+import static java.nio.file.Files.getFileAttributeView;
+import static java.nio.file.Files.newBufferedWriter;
+import static java.nio.file.attribute.PosixFilePermissions.fromString;
+import static java.util.Collections.nCopies;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.joining;
+import static java.util.stream.Collectors.toList;
+import static java.util.stream.StreamSupport.stream;
+import static org.terracotta.build.Utils.mapOf;
+import static org.terracotta.build.plugins.packaging.PackageInternal.UNPACKAGED_JAVA_RUNTIME;
+
+/**
+ * Plugin to support generating Java-based tools with Windows and *nix launch scripts.
+ * <p>
+ * Minimal Example:
+ * <pre>
+ * tool {
+ *   name = 'config-tool'
+ *   mainClass = 'org.terracotta.dynamic_config.cli.config_tool.ConfigTool'
+ * }
+ * </pre>
+ * This tool will generate a complete directory structure for the tool in question including start scripts and all
+ * necessary Java libraries in the configured locations.
+ * Maximal Example:
+ * <pre>
+ * tool {
+ *   name = 'management-server'
+ *   jar = bootJar
+ *   classpath = files([])
+ *   unixTemplate = resources.text.fromFile('src/tool/start.sh.template', 'UTF-8')
+ *   windowsTemplate = resources.text.fromFile('src/tool/start.bat.template', 'UTF-8')
+ *   extraContents {
+ *     into('bin') {
+ *       from project.files('src/tool/legacy/bin')
+ *       eachFile {
+ *         mode 0775
+ *       }
+ *     }
+ *   }
+ * }
+ * </pre>
+ */
+public class ToolPlugin implements Plugin<Project> {
+
+  public static final String TOOL_CATEGORY = "tool";
+
+  private static final String DEFAULT_LIBRARY_PATH = "lib";
+  private static final String DEFAULT_SCRIPT_PATH = "bin";
+
+  private static final URL DEFAULT_UNIX_TEMPLATE = requireNonNull(ToolPlugin.class.getResource("/tool/start.sh.template"));
+  private static final URL DEFAULT_WINDOWS_TEMPLATE = requireNonNull(ToolPlugin.class.getResource("/tool/start.bat.template"));
+
+  @Override
+  public void apply(Project project) {
+    ToolExtension toolExtension = project.getExtensions().create(ToolExtension.class, "tool", ToolExtension.class);
+    toolExtension.getBinDirectory().convention(DEFAULT_SCRIPT_PATH);
+    toolExtension.getLibDirectory().convention(DEFAULT_LIBRARY_PATH);
+    toolExtension.getUnixTemplate().convention(project.getResources().getText().fromUri(DEFAULT_UNIX_TEMPLATE));
+    toolExtension.getWindowsTemplate().convention(project.getResources().getText().fromUri(DEFAULT_WINDOWS_TEMPLATE));
+
+    Project rootProject = project.getRootProject();
+    rootProject.getPlugins().withType(DependencyCheckPlugin.class).configureEach(depCheckPlugin -> {
```

Comment:
I'm thinking this block... which brings in the dependency check dependency should maybe be extracted in to a ToolConvention?

Reply:
yes this was this part which was not compiling.

I can look tomorrow yes if I am able to do it: is it complicated to create this convention class or would required quite a complex refactor ?

Reply:
Should be pretty simple... the pattern already exists for the other convention plugins. You can just follow the existing examples.

Reply:
@chrisdennis : I decided to completely remove the whole block. 

I don't see any dependency with the tool plugin in this code: it just activates owasp dependency check in the root project, right ? So this could be added anywhere else, no need a ToolConvention ?

```java
    Project rootProject = project.getRootProject();
    rootProject.getPlugins().withType(DependencyCheckPlugin.class).configureEach(depCheckPlugin -> {
      rootProject.getExtensions().configure(DependencyCheckExtension.class, dependencyCheck -> {
        dependencyCheck.getScanProjects().add(project.getPath());
      });
    });
```